vortex-ipc/src/reader.rs

```diff
@@ -316,4 +391,138 @@ mod tests {
+            vec![2999989i32, 2999988, 2999987, 2999986, 2899999, 0],
+            PrimitiveEncoding.id(),
+        );
+    }
+
+    #[test]
+    fn test_write_read_alp() {
+        let pdata = PrimitiveArray::from(
+            (0i32..3_000_000)
+                .rev()
+                .map(|v| v as f64 + 0.5)
+                .collect_vec(),
+        )
+        .into_array();
+        let apl_encoded = ALPArray::encode(pdata).unwrap();
+        test_write_read_inner(
+            apl_encoded,
+            vec![
+                2999989.5f64,
+                2999988.5,
+                2999987.5,
+                2999986.5,
+                2899999.5,
+                0.5,
+            ],
+            ALPEncoding.id(),
+        );
+    }
+
+    #[test]
+    fn test_write_read_bitpacked() {
+        let uncompressed = PrimitiveArray::from((0i64..3_000_000).rev().collect_vec());
+
+        let cfg = CompressConfig::new().with_enabled([&BitPackedEncoding as EncodingRef]);
+        let ctx = CompressCtx::new(Arc::new(cfg));
+
+        let packed = BitPackedEncoding {}
+            .compress(uncompressed.array(), None, ctx)
+            .unwrap();
+        assert_eq!(packed.encoding().id(), BitPackedEncoding.id());
+        test_write_read_inner(
+            packed,
+            vec![2999989i64, 2999988, 2999987, 2999986, 2899999, 0],
+            PrimitiveEncoding.id(),
+        );
+    }
+
+    #[test]
+    fn test_write_read_chunked() {
+        let indices = PrimitiveArray::from(vec![10u32, 11, 12, 13, 100_000, 2_999_999, 3_000_000])
+            .into_array();
+        let data = PrimitiveArray::from((0i32..3_000_000).rev().collect_vec()).into_array();
+        let data2 =
+            PrimitiveArray::from((3_000_000i32..6_000_000).rev().collect_vec()).into_array();
+        let chunked = ChunkedArray::try_new(
+            vec![data, data2],
+            DType::Int(IntWidth::_32, Signedness::Signed, Nullability::NonNullable),
+        )
+        .unwrap()
+        .into_array();
+        let mut buffer = vec![];
+        {
+            let mut cursor = Cursor::new(&mut buffer);
+            {
+                let mut writer =
+                    StreamWriter::try_new(&mut cursor, SerdeContext::default()).unwrap();
+                writer.write_array(&chunked).unwrap();
+                writer.write_array(&chunked).unwrap();
+            }
+        }
+
+        let mut cursor = Cursor::new(&buffer);
+        let mut reader = StreamReader::try_new(&mut cursor).unwrap();
+        let mut array_reader = reader.next().unwrap().unwrap();
+        let array = array_reader.take(&indices).unwrap();
+        assert_eq!(array.encoding().id(), ChunkedEncoding.id());
+        let bind = ChunkedArray::try_from(array).unwrap();
+        let mut chunked_out = bind.chunks();
+        assert_eq!(
+            chunked_out
+                .next()
+                .unwrap()
+                .into_primitive()
+                .typed_data::<u32>()
```

Comment:
Huh, this shouldn't work. I think during the refactor we dropped the assertion that `T::PTYPE == array.ptype()` in typed_data and scalar_buffer. Could you add it back, then fix this to be i32?